executor/index_lookup_merge_join.go

```diff
@@ -0,0 +1,686 @@
+			}
+			return false
+		})
+		for i := 0; i < len(rows); i++ {
+			c.AppendRow(rows[i])
+		}
+		task.outerResult = c
+	}
+
+	if omw.filter != nil {
+		var err error
+		task.outerMatch, err = expression.VectorizedFilter(omw.ctx, omw.filter, chunk.NewIterator4Chunk(task.outerResult), task.outerMatch)
+		if err != nil {
+			return task, err
+		}
+		task.memTracker.Consume(int64(cap(task.outerMatch)))
+	}
+	task.outerIter = chunk.NewIterator4Chunk(task.outerResult)
+	return task, nil
+}
+
+func (omw *outerMergeWorker) increaseBatchSize() {
+	if omw.batchSize < omw.maxBatchSize {
+		omw.batchSize *= 2
+	}
+	if omw.batchSize > omw.maxBatchSize {
+		omw.batchSize = omw.maxBatchSize
+	}
+}
+
+func (imw *innerMergeWorker) run(ctx context.Context, wg *sync.WaitGroup) {
+	var task *lookUpMergeJoinTask
+	defer func() {
+		if r := recover(); r != nil {
+			err := errors.Errorf("%v", r)
+			logutil.Logger(context.Background()).Error("innerMergeWorker panic in the recoverable goroutine", zap.Error(err))
+			task.doneErr = make(chan error, 1)
+			task.doneErr <- err
+		}
+		wg.Done()
+	}()
+
+	for ok := true; ok; {
+		select {
+		case task, ok = <-imw.taskCh:
+			if !ok {
+				return
+			}
+		case <-ctx.Done():
+			return
+		}
+
+		err := imw.handleTask(ctx, task)
+		task.doneErr <- err
+	}
+}
+
+func (imw *innerMergeWorker) handleTask(ctx context.Context, task *lookUpMergeJoinTask) error {
+	dLookUpKeys, err := imw.constructDatumLookupKeys(task)
+	if err != nil {
+		return err
+	}
+	dLookUpKeys = imw.dedupDatumLookUpKeys(dLookUpKeys)
+	imw.innerExec, err = imw.readerBuilder.buildExecutorForIndexJoin(ctx, dLookUpKeys, imw.indexRanges, imw.keyOff2IdxOff, imw.nextColCompareFilters)
+	if err != nil {
+		return err
+	}
+	defer terror.Call(imw.innerExec.Close)
+	_, err = imw.fetchInnerResult(ctx, task)
+	if err != nil {
+		return err
+	}
+	err = imw.doMergeJoin(ctx, task)
+	return err
+}
+
+func (imw *innerMergeWorker) checkChunkIsFull(ctx context.Context, task *lookUpMergeJoinTask, chk **chunk.Chunk) bool {
+	if (*chk).IsFull() {
+		select {
+		case task.results <- *chk:
+		case <-ctx.Done():
+			return true
+		}
+		*chk = chunk.NewChunkWithCapacity(imw.retFieldTypes, imw.maxChunkSize)
+	}
+	return false
+}
+
+func (imw *innerMergeWorker) doMergeJoin(ctx context.Context, task *lookUpMergeJoinTask) (err error) {
+	chk := chunk.NewChunkWithCapacity(imw.retFieldTypes, imw.maxChunkSize)
+	defer func() {
+		if chk.NumRows() > 0 {
+			select {
+			case task.results <- chk:
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	if task.innerResult == nil || task.innerResult.NumRows() == 0 {
+		for outerRow := task.outerIter.Begin(); outerRow != task.outerIter.End(); outerRow = task.outerIter.Next() {
+			imw.joiner.onMissMatch(false, outerRow, chk)
+			if imw.checkChunkIsFull(ctx, task, &chk) {
+				return nil
+			}
+		}
+		return
+	}
+
+	task.sameKeyIter = chunk.NewIterator4Slice(task.sameKeyInnerRows)
```

Comment:
useless line.

Reply:
It's useful... We should initialize the `sameKeyIter` to avoid panic.

Reply:
But `sameKeyIter` can be promised to have been initiated in `fetchSameKeyInnerRows` ?

Reply:
Oh, that's right.